### PR TITLE
Introduce SendEmptyResponse method in CoAP Server.

### PR DESCRIPTION
**Pull request by LuDuda**
This PR unifies a way of how CoAP Ntf/Qry/AnsCon responses are handled.

### CODE REVIEW

File: src/core/thread/announce_begin_server.cpp [hunk 1]
```diff
@@ -118,40 +119,13 @@ void AnnounceBeginServer::HandleRequest(Coap::Header &aHeader, Message &aMessage
...
-}
-
-ThreadError AnnounceBeginServer::SendResponse(const Coap::Header &aRequestHeader, const Ip6::MessageInfo &aRequestInfo)
-{
-    ThreadError error = kThreadError_None;
-    Message *message = NULL;
-    Coap::Header responseHeader;
-    Ip6::MessageInfo responseInfo(aRequestInfo);
-
-    VerifyOrExit(aRequestHeader.GetType() == kCoapTypeConfirmable, ;);
```
Comment (nibanks) [inline src/core/thread/announce_begin_server.cpp]: I noticed this isn't in `SendEmtpyResponse`. Should it?

Comment (LuDuda) [inline src/core/thread/announce_begin_server.cpp]: It should, great catch! Will add it.

File: src/core/thread/address_resolver.cpp [hunk 2]
```diff
@@ -281,7 +281,11 @@ void AddressResolver::HandleAddressNotification(Coap::Header &aHeader, Message &
             mCache[i].mTimeout = 0;
             mCache[i].mFailures = 0;
             mCache[i].mState = Cache::kStateCached;
-            SendAddressNotificationResponse(aHeader, aMessageInfo);
+
+            SuccessOrExit(mCoapServer.SendEmptyResponse(aHeader, aMessageInfo));
```
Comment (jwhui) [inline src/core/thread/address_resolver.cpp]: I think we still want to call `HandleResolved()` below, even if we failed to send a response.

File: src/core/thread/address_resolver.cpp [hunk 3]
```diff
@@ -415,7 +367,9 @@ void AddressResolver::HandleAddressError(Coap::Header &aHeader, Message &aMessag
 
     if (!aMessageInfo.GetSockAddr().IsMulticast())
     {
-        SendAddressErrorResponse(aHeader, aMessageInfo);
+        SuccessOrExit(mCoapServer.SendEmptyResponse(aHeader, aMessageInfo));
```
Comment (jwhui) [inline src/core/thread/address_resolver.cpp]: I think we still want to process this Address Error message, even if we failed to send a response.

File: src/core/thread/network_diagnostic.cpp [hunk 4]
```diff
@@ -471,7 +444,7 @@ void NetworkDiagnostic::HandleDiagnosticGetQuery(Coap::Header &aHeader, Message
     // DIAG_GET.qry may be sent as a confirmable message.
     if (aHeader.GetType() == kCoapTypeConfirmable)
     {
-        SuccessOrExit(error = SendEmptyAck(aHeader, aMessageInfo));
+        SuccessOrExit(mCoapServer.SendEmptyResponse(aHeader, aMessageInfo));
```
Comment (jwhui) [inline src/core/thread/network_diagnostic.cpp]: I think we still want to process the Query message, even if sending the response fails.

**Comment by jwhui** [comment]
@LuDuda, can you rebase on master to get the build checks to pass?  Thanks.

File: src/core/thread/announce_begin_server.cpp (hunk 1 above)
Comment (bukepo) [inline src/core/thread/announce_begin_server.cpp]: How about putting this check inside `SendEmptyResponse()`?

File: src/core/thread/energy_scan_server.cpp [hunk 5]
```diff
@@ -104,40 +105,14 @@ void EnergyScanServer::HandleRequest(Coap::Header &aHeader, Message &aMessage, c
...
-}
-
-ThreadError EnergyScanServer::SendResponse(const Coap::Header &aRequestHeader, const Ip6::MessageInfo &aRequestInfo)
-{
-    ThreadError error = kThreadError_None;
-    Message *message = NULL;
-    Coap::Header responseHeader;
-    Ip6::MessageInfo responseInfo(aRequestInfo);
-
-    VerifyOrExit(aRequestHeader.GetType() == kCoapTypeConfirmable, ;);
```
Comment (librasungirl) [inline src/core/thread/energy_scan_server.cpp]: Here also should check type before `SendEmptyResponse`

**Comment by LuDuda** [comment]
Thanks for feedback! So the changes are:
 - Change name to SendEmptyAck.
 - Using "if-statement" instead of SuccessOrExit in case we want to process the request further. But i can remove SuccessOrExit completely if you want?
 - Put verify if header is of `kCoapTypeConfirmable `type to `SendEmptyAck` as @bukepo suggested. Imho it makes code less errorprone but it hides some logic which i think should be handled in layers above CoAP.

**Comment by LuDuda** [comment]
Shall we accept this PR or any changes/improvements needed from my side?

**Pull request merged**